src/meta/processors/admin/BalanceTask.cpp

```diff
@@ -51,10 +52,13 @@ void BalanceTask::invoke() {
             SAVE_STATE();
             if (srcLived_) {
                 client_->transLeader(spaceId_, partId_, src_).thenValue([this](auto&& resp) {
-                    if (!resp.ok()) {
-                        ret_ = Result::FAILED;
-                    } else {
-                        status_ = Status::ADD_PART_ON_DST;
+                    {
+                        std::lock_guard<std::mutex> lg(lock_);
```

Comment:
这里举一个简单的例子，假设现在只有一个cpu，两个线程，当前线程调用client_->transLeader后，操作系统调度client的线程执行，这样client_->transLeader返回的future的状态就是ready的了，此时再调度到当前线程，由于future的状态是ready的，就会直接调用thenValue的cb，这样调用关系就变成了：
BalanceTask::invoke():
    lock(lock_)
    ....
    cb():
        lock(lock_)
     ....

Reply:
就算是这个例子也不会死锁吧……没有循环依赖啊……

Reply:
这个lock_是mutex，不支持递归加锁，第一个lock已经锁住了，由于cb是在当前线程栈中直接调用的，所以invoke中的lock不会析构释放锁，这样在cb中再尝试加锁的时候就死锁了

Reply:
你说的这个是cb在同一个线程，到底几个线程？

Reply:
我终于明白你意思了，你说的是当promise.setValue时，thenValue会被直接调用对吧？你说的是类似下面这个情况，这是会出现卡死的情况的，但是和这个是不同的……
```
void invoke() {
    std::lock_guard<std::mutex> guard(lock);
    LOG(INFO) << "In invoke " << x;
    response().thenValue([this] (auto&& resp) {
        LOG(INFO) << "In thenValue " << x;
        std::lock_guard<std::mutex> lg(lock);
        if (resp.ok()) {
            x = 3;
        }
    });
    x = 1;
    LOG(INFO) << "In invoke " << x;
}

folly::Future<Status> response() {
    auto* evb = ioThreadPool_->getEventBase();
    folly::Promise<Status> pro;
    auto f = pro.getFuture();
    LOG(INFO) << "Before via " << x;
    folly::via(evb, [evb, p = std::move(pro), this] () mutable {
        LOG(INFO) << "In via " << x;
        std::lock_guard<std::mutex> guard(lock);
        x = 2;
        p.setValue(Status::OK());
    });
    return f;
}
```

Reply:
哈哈，是的，我说的就是这个意思：当promise.setValue时，thenValue会被直接调用。
而且你举得例子非常正确。
关键balance的代码和你举得例子是一样的，因为balancer是在自己的线程中执行的，看一下balancer的构造函数：
`Balancer(kvstore::KVStore* kv, std::unique_ptr<AdminClient> client)
        : kv_(kv)
        , client_(std::move(client)) {
        executor_.reset(new folly::CPUThreadPoolExecutor(1));
    }`
和balance函数：
`StatusOr<BalanceID> Balancer::balance() {
    bool expected = false;
    if (running_.compare_exchange_strong(expected, true)) {
        if (!recovery()) {
            LOG(ERROR) << "Recovery balancer failed!";
            return Status::Error("Can't do balance because there is one corruptted balance plan!");
        }
        if (plan_ == nullptr) {
            LOG(INFO) << "There is no corrupted plan need to recovery, so create a new one";
            auto status = buildBalancePlan();
            if (plan_ == nullptr) {
                LOG(ERROR) << "Create balance plan failed, status " << status.toString();
                return status;
            }
        }
        **executor_->add(std::bind(&BalancePlan::invoke, plan_.get()));**
        return plan_->id();
    }
    CHECK(!!plan_);
    return plan_->id();
}`
BalancePlan::invoke是在executor_线程中执行的，而client_->transLeader是在全局的ioThreadPool中执行的，所以和你举得例子是一样的

Reply:
哈哈，是的，我说的就是这个意思：当promise.setValue时，thenValue会被直接调用。
而且你举得例子非常正确。
关键balance的代码和你举得例子是一样的，因为balancer是在自己的线程中执行的，看一下balancer的构造函数：
`Balancer(kvstore::KVStore* kv, std::unique_ptr<AdminClient> client)
        : kv_(kv)
        , client_(std::move(client)) {
        executor_.reset(new folly::CPUThreadPoolExecutor(1));
    }`
和balance函数：
`StatusOr<BalanceID> Balancer::balance() {
    bool expected = false;
    if (running_.compare_exchange_strong(expected, true)) {
        if (!recovery()) {
            LOG(ERROR) << "Recovery balancer failed!";
            return Status::Error("Can't do balance because there is one corruptted balance plan!");
        }
        if (plan_ == nullptr) {
            LOG(INFO) << "There is no corrupted plan need to recovery, so create a new one";
            auto status = buildBalancePlan();
            if (plan_ == nullptr) {
                LOG(ERROR) << "Create balance plan failed, status " << status.toString();
                return status;
            }
        }
        **executor_->add(std::bind(&BalancePlan::invoke, plan_.get()));**
        return plan_->id();
    }
    CHECK(!!plan_);
    return plan_->id();
}`
BalancePlan::invoke是在executor_线程中执行的，而client_->transLeader是在全局的ioThreadPool中执行的，所以和你举得例子是一样的

Reply:
client_的方法不是全局的ioThreadPool，全局的是通过`getCPUExecutor`和`getIOExecutor `获取的。Balancer里executor_和AdminClient的ioThreadPoll中取的线程是两个，所以是不会出现卡死的。